### PR TITLE
[CDAP-7154] Add Timeout to Zookeeper Client in Master and Upgrade Tool

### PR DESCRIPTION
**Pull request by knazir**
Issue: https://issues.cask.co/browse/CDAP-7154
Build: http://builds.cask.co/browse/CDAP-DUT4720-1

### CODE REVIEW

File: cdap-common/src/main/resources/cdap-default.xml [hunk 1]
```diff
@@ -200,6 +200,13 @@
     </description>
   </property>
 
+  <property>
+    <name>zookeeper.client.timeout.millis</name>
+    <value>40000</value>
+    <description>
+      ZooKeeper client timeout in milliseconds
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: What do mean by client timeout? It's a bit unclear to me.

Comment (knazir) [reply cdap-common/src/main/resources/cdap-default.xml]: Would 

> Maximum amount of time in milliseconds to wait for the ZooKeeper client to connect to the quorum.

be better? Or maybe 

> Maximum amount of time in milliseconds to wait for ZooKeeper to connect to the quorum on startup.

Comment (chtyim) [reply cdap-common/src/main/resources/cdap-default.xml]: It is not entirely true. It is more like the max time to wait for the first successful connection to one of the server in a ZK quorum.

File: cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java [hunk 2]
```diff
@@ -221,15 +222,22 @@ public void init(String[] args) {
...
     createDirectory(cConf.get(QueueConstants.ConfigKeys.QUEUE_TABLE_COPROCESSOR_DIR,
                               QueueConstants.DEFAULT_QUEUE_TABLE_COPROCESSOR_DIR));
     createSystemHBaseNamespace();
     updateConfigurationTable();
 
-    zkClient.startAndWait();
+    ListenableFuture<Service.State> startFunction = zkClient.start();
+    try {
+      startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java]: we should log a different message based on the type of exception. If its a TimeoutException, we want to log that it timed out and that the zk quorum setting should be checked for correctness.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 3]
```diff
@@ -200,6 +200,14 @@
     </description>
   </property>
 
+  <property>
+    <name>zookeeper.client.timeout.millis</name>
```
Comment (albertshau) [inline cdap-common/src/main/resources/cdap-default.xml]: think it would be better to name this something like 'zookeeper.client.startup.timeout.millis' or something like that. Otherwise if I look at just the name and not the description (which is normally what happens in cdap-site.xml), I would probably think it means something else.

File: cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java [hunk 4]
```diff
@@ -221,15 +223,29 @@ public void init(String[] args) {
...
     createSystemHBaseNamespace();
     updateConfigurationTable();
 
-    zkClient.startAndWait();
+    ListenableFuture<Service.State> startFunction = zkClient.start();
+    try {
+      startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      LOG.error("Connection timed out while trying to start ZooKeeper client. Please verify that the ZooKeeper " +
+                "quorum settings are correct.", e);
```
Comment (poornachandra) [inline cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java]: It would be good to print the quorum setting currently being used.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 5]
```diff
@@ -200,6 +200,14 @@
     </description>
   </property>
 
+  <property>
+    <name>zookeeper.client.startup.timeout.millis</name>
+    <value>600000</value>
```
Comment (poornachandra) [inline cdap-common/src/main/resources/cdap-default.xml]: Do we need to wait for 10 mins before timing out?

**Comment by poornachandra** [comment]
@knazir It would be good to check how auth server, router and cdap-kafka behave when quorum is incorrect.

**Comment by poornachandra** [comment]
I also see lot of code duplication while waiting for connection, and handling timeouts. Would it be possible to refactor that bit out?

File: cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java [hunk 6]
```diff
@@ -250,8 +256,24 @@ public DatasetFramework getInDsFramework(DatasetFramework dsFramework) {
...
   private void startUp() throws Exception {
     // Start all the services.
-    zkClientService.startAndWait();
+    LOG.info("Starting Zookeeper Client...");
+    ListenableFuture<Service.State> startFunction = zkClientService.start();
+    try {
+      startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      LOG.error("Connection timed out while trying to start ZooKeeper client. Please verify that the ZooKeeper " +
+                  "quorum settings are correct.", e);
```
Comment (awholegunch) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java]: See @poornachandra comment above about printing the current settings.

File: cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java [hunk 7]
```diff
@@ -248,7 +252,20 @@ public void startUp() throws Exception {
     }
 
     if (zkClient != null) {
-      zkClient.startAndWait();
+      ListenableFuture<Service.State> startFunction = zkClient.start();
+      try {
+        startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+      } catch (TimeoutException e) {
+        LOG.error("Connection timed out while trying to start ZooKeeper client. Please verify that the ZooKeeper " +
```
Comment (awholegunch) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: ibid.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 8]
```diff
@@ -92,6 +92,7 @@
   public static final class Zookeeper {
     public static final String QUORUM = "zookeeper.quorum";
     public static final String CFG_SESSION_TIMEOUT_MILLIS = "zookeeper.session.timeout.millis";
+    public static final String CFG_CLIENT_TIMEOUT_MILLIS = "zookeeper.client.startup.timeout.millis";
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: The constant should be named as 'CLIENT_STARTUP_TIMEOUT_MILLIS`.

File: cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ * Util class used to refactor ZooKeeper functions.
+ */
+public class ZooKeeperUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperUtils.class);
+
+  private ZooKeeperUtils(){}
+
+  public static void connectWithTimeout(ZKClientService zkClientService, CConfiguration cConf) throws Exception {
+    ListenableFuture<Service.State> startFunction = zkClientService.start();
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java]: Better name it as `startFuture`.

File: cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    try {
+      startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      LOG.error("Connection timed out while trying to start ZooKeeper client. Please verify that the ZooKeeper " +
+                  "quorum settings are correct. Currently configured as: {}", cConf.get(Constants.Zookeeper.QUORUM), e);
+      throw e;
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java]: You have to stop the `zkClientService` if it is timeout or interrupted.

File: cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Util class used to refactor ZooKeeper functions.
+ */
+public class ZooKeeperUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperUtils.class);
+
+  private ZooKeeperUtils(){}
+
+  public static void connectWithTimeout(ZKClientService zkClientService, CConfiguration cConf) throws Exception {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZooKeeperUtils.java]: I don't like this util method (and this class). It's too ZK specific. If It's better to have a util class for `Service` (call it `Services`) and have a method like this:

``` java
public static void startAndWait(Service service, long timeout, TimeUnit timeoutUnit) throws TimeoutException, InterruptedException {
  try {
    service.start().get(timeout, timeoutUnit);
  } catch (TimeoutException e) {
    TimeoutException ex = new TimeoutException(String.format("Service '%s' failed to start in %d %s", service, timeout, timeoutUnit));
    ex.addSuppressed(e);
    service.stop();
    throw ex;
  } catch (InterruptedException e) {
    service.stop();
    throw e;
  } catch (ExecutionException e) {
    throw Throwables.propagate(e.getCause());
  }
}
```

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * Utility class for handling {@link Service} objects.
+ */
+public class Services {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Services.class);
+
+  private Services(){}
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/service/Services.java:36]: Does it pass checkstyle?

File: cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java [hunk 13]
```diff
@@ -86,6 +87,7 @@
   @Inject
   DatasetServiceManager(CConfiguration cConf, Configuration hConf) {
     Injector injector = createInjector(cConf, hConf);
+    this.cConf = cConf;
```
Comment (chtyim) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java]: What is this for?

Comment (knazir) [reply cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java]: Forgot to revert this change when I removed the timeout in `DatasetServiceManager`, oops. Thanks.

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java (hunk 12 above)
Comment (knazir) [reply cdap-common/src/main/java/co/cask/cdap/common/service/Services.java:36]: Yep

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+   * @throws ExecutionException If an exception occurs while trying to start the service
+   */
+  public static void startAndWait(Service service, long timeout, TimeUnit timeoutUnit,
+                                  @Nullable String timeoutErrorMessage) throws Exception {
+    ListenableFuture<Service.State> startFuture = service.start();
+    try {
+      startFuture.get(timeout, timeoutUnit);
+    } catch (TimeoutException e) {
+      LOG.error(timeoutErrorMessage != null ? timeoutErrorMessage : "Timeout while waiting to start service.", e);
+      service.stop();
```
Comment (poornachandra) [inline cdap-common/src/main/java/co/cask/cdap/common/service/Services.java]: It would be good to wrap the stop in try-catch. Otherwise, if stop throws exception then the original exception will be lost.

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+      throw e;
+    } catch (InterruptedException e) {
+      LOG.error("Interrupted while waiting to start service.", e);
+      service.stop();
```
Comment (poornachandra) [inline cdap-common/src/main/java/co/cask/cdap/common/service/Services.java]: Same here

File: cdap-common/src/main/resources/cdap-default.xml [hunk 16]
```diff
@@ -200,6 +200,14 @@
     </description>
   </property>
 
+  <property>
+    <name>zookeeper.client.startup.timeout.millis</name>
+    <value>60000</value>
+    <description>
+      Maximum amount of time in milliseconds to wait for the first successful
```
Comment (poornachandra) [inline cdap-common/src/main/resources/cdap-default.xml]: `first successful`? Can it be just `successful`?

**Comment by knazir** [comment]
This is the exception shown in master, kafka, and router logs:

```
Exception in thread "main" java.util.concurrent.TimeoutException: Connection timed out while trying to start ZooKeeper client. Please verify that the ZooKeeper quorum settings are correct. Currently configured as: 219.88.235.124:2941/cdap
        at com.google.common.util.concurrent.AbstractService$Transition.get(AbstractService.java:355)
        at com.google.common.util.concurrent.AbstractService$Transition.get(AbstractService.java:348)
        at co.cask.cdap.common.service.Services.startAndWait(Services.java:52)
        at co.cask.cdap.data.runtime.main.MasterServiceMain.start(MasterServiceMain.java:232)
        at co.cask.cdap.common.runtime.DaemonMain.doMain(DaemonMain.java:58)
        at co.cask.cdap.data.runtime.main.MasterServiceMain.main(MasterServiceMain.java:179)
```

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+   * @param service The service to start
+   * @param timeout The duration to wait for the service to start
+   * @param timeoutUnit The time unit used for the timeout parameter
+   * @param timeoutErrorMessage An optional error message to display if starting the service times out
+   * @throws TimeoutException If the service can not be started before the specified timeout
+   * @throws InterruptedException If the service is interrupted while trying to start the service
+   * @throws ExecutionException If an exception occurs while trying to start the service
+   */
+  public static void startAndWait(Service service, long timeout, TimeUnit timeoutUnit,
+                                  @Nullable String timeoutErrorMessage) throws Exception {
```
Comment (poornachandra) [inline cdap-common/src/main/java/co/cask/cdap/common/service/Services.java]: It is good to list the exceptions instead of the generic `Exception` in throws clause.

File: cdap-common/src/main/java/co/cask/cdap/common/service/Services.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    } catch (InterruptedException e) {
+      LOG.error("Interrupted while waiting to start service.", e);
+      try {
+        service.stop();
+      } catch (Exception stopException) {
+        LOG.error("Error while trying to stop service:", stopException);
+      }
+      throw e;
+    } catch (ExecutionException e) {
+      LOG.error("Exception while trying to start service.", e);
```
Comment (poornachandra) [inline cdap-common/src/main/java/co/cask/cdap/common/service/Services.java]: I think this log may not be required

File: cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java [hunk 19]
```diff
@@ -76,7 +76,7 @@
  * Note: This should not be used outside upgrade tool.
  */
 public class DatasetServiceManager extends AbstractIdleService {
-
+
```
Comment (poornachandra) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetServiceManager.java]: This change looks unrelated to this PR

**Comment by poornachandra** [comment]
LGTM, please squash and merge after build.

**Comment by knazir** [comment]
Guild is breen: http://builds.cask.co/browse/CDAP-DUT4720-4 going to squash and merge.
